activerecord/lib/active_record/autosave_association.rb

```diff
@@ -505,8 +505,14 @@ def save_belongs_to_association(reflection)
             saved = record.save(validate: !autosave) if record.new_record? || (autosave && record.changed_for_autosave?)
 
             if association.updated?
-              association_id = record.public_send(reflection.options[:primary_key] || :id)
-              self[reflection.foreign_key] = association_id unless self[reflection.foreign_key] == association_id
+              primary_key = Array(reflection.options[:primary_key] || reflection.active_record_primary_key)
```

Comment:
I want to specifically highlight the `:id` => `reflection.active_record_primary_key` change. I believe usage of `:id` here was a little confusing as it may appear like we are hardcoding the "id column value" in this context while actually `:id` here is being used as a method call which by implication returns the `@primary_key` value. So similar change could have happened even out of the scope of this PR. 

In scope of this PR the change has to happen as we transition to a higher-level `query_constraints_list` abstraction which in this case may be treated as a "virtual primary key" and for all existing models it equals to `[:id]` . And we have following options for making the change:
- Ask the record for columns names like `record.class.query_constraints_list`. I don't like this option as it couples `AutosaveAssociation` module with lower-level `query_constraints` model concept.
- Ask `reflection` for the primary key value. This is my preferable choice as `reflection` is what actually defines how associations should behave in a given context.

Reply:
I end up using `record.class.query_constraints_list` as asking `reflection` for the value breaks a few corner-cases which I missed. I still feel like that the value should be derived through `reflection` and not directly from the `record` though I'm not familiar with the associations internals enough to be confident. Using `record.class.query_constraints_list` narrows down the scope of the change and avoid breaking these corner-cases